In [1]:
import pandas as pd

import numpy as np

import matplotlib.pyplot as plt

import seaborn as sns

In [2]:
df = pd.read_csv("Retail Data.csv")
df

Order No  Order Date           Customer Name  \
0      4293-1  02-09-2014        Vivek Sundaresam   
1      5001-1  24-10-2015          Shahid Hopkins   
2      5004-1  13-03-2014           Dennis Pardue   
3      5009-1  18-02-2013              Sean Wendt   
4      5010-1  13-09-2014  Christina Vanderzanden   
...       ...         ...                     ...   
4995   6831-1  02-11-2016           Roland Murray   
4996   6847-1  16-12-2015          Patrick OBrill   
4997   6922-1  28-09-2016            Troy Staebel   
4998   6939-1  18-06-2016             Shaun Weien   
4999   6980-1  12-08-2015          Nora Pelletier   

                                  Address       City State   Customer Type  \
0          152 Bunnerong Road,Eastgardens     Sydney   NSW  Small Business   
1           438 Victoria Avenue,Chatswood     Sydney   NSW       Corporate   
2                412 Brunswick St,Fitzroy  Melbourne   VIC        Consumer   
3                145 Ramsay St,Haberfield     Sydney   NSW  Small Business   
4                  188 Pitt Street,Sydney     Sydney   NSW  Small Business   
...                                   ...        ...   ...             ...   
4995        438 Victoria Avenue,Chatswood     Sydney   NSW       Corporate   
4996        63 Riley Street,Woolloomooloo     Sydney   NSW  Small Business   
4997             1-2/299 Sussex St,Sydney     Sydney   NSW        Consumer   
4998          Shop 3/144 Wattle St,Ultimo     Sydney   NSW       Corporate   
4999  5/250 Old Northern Road,Castle Hill     Sydney   NSW        Consumer   

         Account Manager Order Priority  \
0           Tina Carlton       Critical   
1           Natasha Song         Medium   
2           Connor Betts  Not Specified   
3            Phoebe Gour       Critical   
4           Tina Carlton  Not Specified   
...                  ...            ...   
4995        Natasha Song       Critical   
4996        Tina Carlton  Not Specified   
4997         Aanya Zhang            Low   
4998     Samantha Chairs  Not Specified   
4999  Nicholas Fernandes            Low   

                                        Product Name  ... Cost Price  \
0     UGen Ultra Professional Cordless Optical Suite  ...    $156.50   
1                                Bagged Rubber Bands  ...      $0.24   
2                    TechSavi Cordless Navigator Duo  ...     $42.11   
3      Artisan Printable Repositionable Plastic Tabs  ...      $5.33   
4                          Pizazz Drawing Pencil Set  ...      $1.53   
...                                              ...  ...        ...   
4995        Artisan Hi-Liter Smear-Safe Highlighters  ...      $2.98   
4996            Smiths Colored Interoffice Envelopes  ...     $19.83   
4997                             Laser DVD-RAM discs  ...     $20.18   
4998                              Artisan 479 Labels  ...      $1.59   
4999                                    Binder Posts  ...      $3.50   

     Retail Price Profit Margin Order Quantity  Sub Total Discount %  \
0         $300.97       $144.47           23.0  $4,533.52         2%   
1           $1.26         $1.02            8.0     $45.20         3%   
2          $80.98        $38.87           45.0    $873.32         4%   
3           $8.60         $3.27           16.0     $73.52         1%   
4           $2.78         $1.25           49.0    $138.46         7%   
...           ...           ...            ...        ...        ...   
4995        $5.84         $2.86           35.0    $115.40         2%   
4996       $30.98        $11.15           49.0  $1,999.69         7%   
4997       $35.41        $15.23           19.0    $929.40         9%   
4998        $2.61         $1.02           23.0     $66.54         5%   
4999        $5.74         $2.24           31.0    $213.34         0%   

     Discount $  Order Total Shipping Cost      Total  
0       $194.83    $4,757.22         $7.18  $4,291.55  
1         $0.00       $45.90         $0.70     $46.91  
2        $72.23 

In [3]:
#01.  View the structure of the dataset (columns, types, missing values). 

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 24 columns):
 #   Column             Non-Null Count  Dtype  
---  ------             --------------  -----  
 0   Order No           5000 non-null   object 
 1   Order Date         5000 non-null   object 
 2   Customer Name      5000 non-null   object 
 3   Address            4999 non-null   object 
 4   City               5000 non-null   object 
 5   State              5000 non-null   object 
 6   Customer Type      5000 non-null   object 
 7   Account Manager    5000 non-null   object 
 8   Order Priority     5000 non-null   object 
 9   Product Name       5000 non-null   object 
 10  Product Category   5000 non-null   object 
 11  Product Container  5000 non-null   object 
 12  Ship Mode          5000 non-null   object 
 13  Ship Date          5000 non-null   object 
 14  Cost Price         5000 non-null   object 
 15  Retail Price       5000 non-null   object 
 16  Profit Margin      5000 

In [4]:
#02. What is the shape (rows, columns) of the dataset? 

df.shape

(5000, 24)

In [5]:
#03. Are there any duplicate records?\

df.duplicated().sum()

np.int64(0)

In [6]:
#04. Are there any missing or corrupted entries in Ship Date, Order Date, or numeric columns?

columns = ["Order Date","Ship Date","Cost Price","Retail Price","Profit Margin","Order Quantity","Sub Total","Discount %","Order Total","Shipping Cost","Total"]

df[columns].isnull().sum()

Order Date        0
Ship Date         0
Cost Price        0
Retail Price      0
Profit Margin     0
Order Quantity    1
Sub Total         0
Discount %        0
Order Total       0
Shipping Cost     0
Total             0
dtype: int64

In [7]:
# #05. Convert Order Date and Ship Date to datetime. 

df["Order Date"] = pd.to_datetime(df["Order Date"],errors='coerce',dayfirst=True)

df["Ship Date"] = pd.to_datetime(df["Ship Date"],errors='coerce',dayfirst=True)

df

print(df["Order Date"].dtype)
print(df["Ship Date"].dtype)



datetime64[ns]
datetime64[ns]


In [8]:
#06. Check for future or inconsistent shipping dates. 

current_date = pd.to_datetime('now')

#checking for Future_shipping date
future_date = df[df["Ship Date"] > current_date] 

print("---Future dates----")
if not future_date.empty:
    print(f"Found {len(future_date)} entries with future date")
    print(future_date["Order No"],["Order Date"],["Ship date"])
else:
    print("No Future dates found")

# checking for Inconsistent date
Inconsistent_date = df[df["Ship Date"] < current_date] 

if not Inconsistent_date.empty:
    print(f"Found {len(Inconsistent_date)} entries with future date")
    print(Inconsistent_date["Order No"],["Order Date"],["Ship date"])
else:
    print("No Inconsistent date found")

---Future dates----
No Future dates found
Found 5000 entries with future date
0       4293-1
1       5001-1
2       5004-1
3       5009-1
4       5010-1
         ...  
4995    6831-1
4996    6847-1
4997    6922-1
4998    6939-1
4999    6980-1
Name: Order No, Length: 5000, dtype: object ['Order Date'] ['Ship date']


In [9]:
#07. Convert price columns to numeric (remove $ and commas). 

price_column = ["Cost Price","Retail Price"]

for col in price_column:
    df[col]=df[col].str.replace('$', " ",regex=False)
    df[col]=df[col].str.replace(','," ",regex=False)
    df[col] = pd.to_numeric(df[col],errors="coerce")
print(df[price_column].dtypes)

Cost Price      float64
Retail Price    float64
dtype: object


In [10]:
#08.What are the unique values in Customer Type and Order Priority? 


ct_unique =df["Customer Type"].unique()
print(f"Unique values in customer types are:\n{ct_unique}")

op_unique =df["Order Priority"].unique()
print(f"Unique values in customer types are:\n{op_unique}")




Unique values in customer types are:
['Small Business' 'Corporate' 'Consumer' 'Home Office']
Unique values in customer types are:
['Critical' 'Medium' 'Not Specified' 'Low' 'High']


In [11]:
#09.what are the most common shipping modes? 

common_mode =df["Ship Mode"].max()
print(f"Most common mode of shippinf is :- {common_mode}")

Most common mode of shippinf is :- Regular Air


In [12]:
#10.  Which cities have the highest number of orders? 

cities = df.groupby("City")["Order No"].count().sort_values(ascending=False)
cities

City
Sydney       3584
Melbourne    1416
Name: Order No, dtype: int64

In [ ]:
#11. Whats the range of order quantities and prices? 

# Range of order quantities 
max_order = df["Order Quantity"].max()
min_order = df["Order Quantity"].min()

print(f" Range of Order Quantities is in between {min_order} to {max_order}")

# Range of order prices
max_Price = df["Cost Price"].max()
min_Price= df["Cost Price"].min()

print(f" Range of Price is in between {min_Price} to {max_Price}")


 Range of Order Quantities is in between 1.0 to 50.0
 Range of Price is in between 0.24 to 377.99


In [14]:
#12. Create a new column for shipping duration. 

df["Shipping_duration"]=df["Ship Date"]-df["Order Date"]

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5000 entries, 0 to 4999
Data columns (total 25 columns):
 #   Column             Non-Null Count  Dtype          
---  ------             --------------  -----          
 0   Order No           5000 non-null   object         
 1   Order Date         5000 non-null   datetime64[ns] 
 2   Customer Name      5000 non-null   object         
 3   Address            4999 non-null   object         
 4   City               5000 non-null   object         
 5   State              5000 non-null   object         
 6   Customer Type      5000 non-null   object         
 7   Account Manager    5000 non-null   object         
 8   Order Priority     5000 non-null   object         
 9   Product Name       5000 non-null   object         
 10  Product Category   5000 non-null   object         
 11  Product Container  5000 non-null   object         
 12  Ship Mode          5000 non-null   object         
 13  Ship Date          5000 non-null   datetime64[ns

In [15]:
df

Order No Order Date           Customer Name  \
0      4293-1 2014-09-02        Vivek Sundaresam   
1      5001-1 2015-10-24          Shahid Hopkins   
2      5004-1 2014-03-13           Dennis Pardue   
3      5009-1 2013-02-18              Sean Wendt   
4      5010-1 2014-09-13  Christina Vanderzanden   
...       ...        ...                     ...   
4995   6831-1 2016-11-02           Roland Murray   
4996   6847-1 2015-12-16          Patrick OBrill   
4997   6922-1 2016-09-28            Troy Staebel   
4998   6939-1 2016-06-18             Shaun Weien   
4999   6980-1 2015-08-12          Nora Pelletier   

                                  Address       City State   Customer Type  \
0          152 Bunnerong Road,Eastgardens     Sydney   NSW  Small Business   
1           438 Victoria Avenue,Chatswood     Sydney   NSW       Corporate   
2                412 Brunswick St,Fitzroy  Melbourne   VIC        Consumer   
3                145 Ramsay St,Haberfield     Sydney   NSW  Small Business   
4                  188 Pitt Street,Sydney     Sydney   NSW  Small Business   
...                                   ...        ...   ...             ...   
4995        438 Victoria Avenue,Chatswood     Sydney   NSW       Corporate   
4996        63 Riley Street,Woolloomooloo     Sydney   NSW  Small Business   
4997             1-2/299 Sussex St,Sydney     Sydney   NSW        Consumer   
4998          Shop 3/144 Wattle St,Ultimo     Sydney   NSW       Corporate   
4999  5/250 Old Northern Road,Castle Hill     Sydney   NSW        Consumer   

         Account Manager Order Priority  \
0           Tina Carlton       Critical   
1           Natasha Song         Medium   
2           Connor Betts  Not Specified   
3            Phoebe Gour       Critical   
4           Tina Carlton  Not Specified   
...                  ...            ...   
4995        Natasha Song       Critical   
4996        Tina Carlton  Not Specified   
4997         Aanya Zhang            Low   
4998     Samantha Chairs  Not Specified   
4999  Nicholas Fernandes            Low   

                                        Product Name  ... Retail Price  \
0     UGen Ultra Professional Cordless Optical Suite  ...       300.97   
1                                Bagged Rubber Bands  ...         1.26   
2                    TechSavi Cordless Navigator Duo  ...        80.98   
3      Artisan Printable Repositionable Plastic Tabs  ...         8.60   
4                          Pizazz Drawing Pencil Set  ...         2.78   
...                                              ...  ...          ...   
4995        Artisan Hi-Liter Smear-Safe Highlighters  ...         5.84   
4996            Smiths Colored Interoffice Envelopes  ...        30.98   
4997                             Laser DVD-RAM discs  ...        35.41   
4998                              Artisan 479 Labels  ...         2.61   
4999                                    Binder Posts  ...         5.74   

     Profit Margin Order Quantity  Sub Total  Discount %  Discount $  \
0          $144.47           23.0  $4,533.52          2%     $194.83   
1            $1.02            8.0     $45.20          3%       $0.00   
2           $38.87           45.0    $873.32          4%      $72.23   
3            $3.27           16.0     $73.52          1%       $4.35   
4            $1.25           49.0    $138.46          7%       $5.95   
...            ...            ...        ...         ...         ...   
4995         $2.86           35.0    $115.40          2%      $10.81   
4996        $11.15           49.0  $1,999.69          7%     $143.91   
4997        $15.23           19.0    $929.40          9%      $27.88   
4998         $1.02           23.0     $66.54          5%       $4.46   
4999         $2.24           31.0    $213.34          0%       $4.67   

     Order Total  Shipping Cost      Total Shipping_duration  
0      $4,757.22          $7.18  $4,291.55            2 days  
1         $45.90          $0.70     $46.91    

In [16]:
#13.  Are there any orders with zero or negative total or quantity? 

df[df["Order Quantity"]<= 0]



Empty DataFrame
Columns: [Order No, Order Date, Customer Name, Address, City, State, Customer Type, Account Manager, Order Priority, Product Name, Product Category, Product Container, Ship Mode, Ship Date, Cost Price, Retail Price, Profit Margin, Order Quantity, Sub Total, Discount %, Discount $, Order Total, Shipping Cost, Total, Shipping_duration]
Index: []

[0 rows x 25 columns]

In [17]:
#14. Are all discount percentages matching discount dollar amounts? 

equal = (df["Discount $"] == df["Discount %"]).all()
equal

np.False_

In [ ]:
#15. Check for mismatches in total calculation.


col = ["Sub Total","Discount $","Shipping Cost","Total"]
for col in col:
    df[col] = (df[col].astype(str).str.strip().str.replace(r'[^0-9.\-]+',"",regex=True))
    df[col] = pd.to_numeric(df[col],errors='coerce')

mis_match_t = df[abs(df["Total"]- (df["Sub Total"]-df["Discount $"]+ df["Shipping Cost"])) > 0.01]

if mis_match_t.empty:
    print("No mismatched total found")
else:
    print(f"miss matched total found in {len(mis_match_t)} rows:")
   

miss matched total found in 3960 rows:


In [19]:
#16. Identify top 5 products by order quantity. 

df.groupby("Product Name")["Order Quantity"].count().sort_values(ascending=False).head(5)

Product Name
Artisan 474 Labels                         136
Smiths General Use 3-Ring Binders          126
Apex Preferred Stainless Steel Scissors    112
Artisan 479 Labels                         112
Alto Memo Cubes                            107
Name: Order Quantity, dtype: int64

In [20]:
#17.  Which Account Manager handled the most revenue? 


df["Revenue"]= df["Retail Price"]*df["Order Quantity"]
total_revenue = df["Revenue"].sum()
df.groupby("Account Manager")["Revenue"].sum().sort_values(ascending=False).head(1)





Account Manager
Connor Betts    928249.39
Name: Revenue, dtype: float64

In [21]:
#18.  What is the average shipping cost by mode? 

df.groupby("Ship Mode")["Shipping Cost"].mean().sort_values(ascending=False)


Ship Mode
Regular Air       5.503553
Delivery Truck    5.270085
Express Air       5.247311
Name: Shipping Cost, dtype: float64

In [22]:
#19.Find the most profitable product. 

df.groupby("Product Name")["Profit Margin"].max().sort_values(ascending=False).head(1)

Product Name
Economy Rollaway Files    $97.47
Name: Profit Margin, dtype: object

Insightful Analysis Questions 

In [23]:
#01. What is the total revenue generated across all orders? 

df["Revenue"]= df["Retail Price"]*df["Order Quantity"]
total_revenue = df["Revenue"].sum()

print(f"Total revenue generated across all orders is :- {total_revenue}")


Total revenue generated across all orders is :- 5186314.12


In [24]:
#02. Which customer type generates more revenue? 

df.groupby("Customer Type")["Revenue"].sum().sort_values(ascending=False).head(1)

Customer Type
Corporate    1753097.41
Name: Revenue, dtype: float64

In [25]:
#03.How does order priority affect revenue? 

df["Order Priority"]=df["Order Priority"].replace(
    {"Not Specified": 0, "Low": 1, "Medium": 2,"High":3,"Critical":4}
)
corr = df["Order Priority"].corr(df["Revenue"])
corr

C:\Users\mjmak\AppData\Local\Temp\ipykernel_13056\3672266194.py:3: FutureWarning: Downcasting behavior in `replace` is deprecated and will be removed in a future version. To retain the old behavior, explicitly call `result.infer_objects(copy=False)`. To opt-in to the future behavior, set `pd.set_option('future.no_silent_downcasting', True)`
  df["Order Priority"]=df["Order Priority"].replace(


np.float64(0.007173550710202885)

In [26]:
#04.What is the average profit margin by product category? 



df["Profit Margin"] = (df["Profit Margin"].astype(str).str.strip().str.replace('$',''))
df["Profit Margin"]= pd.to_numeric(df["Profit Margin"],errors='coerce')

# checking datatype after conversion
df["Profit Margin"].dtype

df["Profit Margin"].isna()

df.groupby("Product Category")["Profit Margin"].mean()





Product Category
Furniture          12.701953
Office Supplies    17.682609
Technology         20.423208
Name: Profit Margin, dtype: float64

In [27]:
#05.What is the most profitable product overall? 

Product_max_profit=df.groupby("Product Name")["Profit Margin"].max().sort_values(ascending=False).head(1)
product = Product_max_profit.index[0]

print(f"Most profitable product is :- {product}")


Most profitable product is :- Emerson Stylus 1520 Color Inkjet Printer


In [28]:
#06. How many days does it usually take to ship an order? 

avg_days =df["Shipping_duration"].mean().days

print(f"Usually it takes {avg_days} day to ship an order")

Usually it takes 1 day to ship an order


In [ ]:
#07.Do longer shipping times impact profit margins?

df["Shipping_duration"].mean()

corr = df["Shipping_duration"].corr(df["Profit Margin"])
corr


np.float64(-0.01978703518675127)

In [39]:
#08.Which city brings in the highest revenue? 

df["Revenue"]= df["Retail Price"]*df["Order Quantity"]

df.groupby("City")["Revenue"].max().sort_values(ascending=False).head(1)


City
Melbourne    29999.5
Name: Revenue, dtype: float64

In [43]:
#09.Which account manager generated the most revenue? 

df.groupby("Account Manager")["Revenue"].max().sort_values(ascending=False).head(1)

Account Manager
Connor Betts    29999.5
Name: Revenue, dtype: float64

In [ ]:
#10. Which shipping mode is most cost-effective (lowest avg. shipping)?

df.groupby("Ship Mode")["Shipping Cost"].mean().sort_values().head(1)




Ship Mode
Express Air    5.247311
Name: Shipping Cost, dtype: float64

In [52]:
#11.  Do higher discounts reduce profits? 

corr = df["Discount $"].corr(df["Profit Margin"])
corr

np.float64(0.39349726173151367)

In [54]:
#12.Which state has the highest number of orders? 

df.groupby("State")["Order Quantity"].count().sort_values(ascending=False).head(1)

State
NSW    3583
Name: Order Quantity, dtype: int64

In [ ]:
#13. What is the average discount % across all orders? 

df["Discount %"] = df["Discount %"].str.strip('%').astype(float)/100

df["Discount %"].mean()

np.float64(0.05033800000000001)

In [60]:
#14. What is the average total spend per order? 

df["Total"].mean()

np.float64(746.3072800000001)

In [71]:
#15. Are certain containers (e.g., Small Box, Wrap Bag) more profitable? 

avg_profit_by_container = df.groupby("Product Container")["Profit Margin"].mean()
avg_profit_by_container

profitable_cont = avg_profit_by_container.idxmax()

containr = avg_profit_by_container.max()
containr


print(f"Most profitable container is {profitable_cont} with profit of {containr:.2f}%")

Most profitable container is Small Box with profit of 19.67%


FINDINGS -->

1.What is the shape (rows, columns) of the dataset? 
--> Shape of dataset is (5000, 24)

2.Are there any duplicate records?
--> There are no duplicate records.

3.Are there any missing or corrupted entries in Ship Date, Order Date, or numeric columns?
--> There is 1 missing/corrupted entry in Order Quantity.

4.What are the unique values in Customer Type and Order Priority? 
--> 

    Unique values in customer types are:
    ['Small Business' 'Corporate' 'Consumer' 'Home Office']

    Unique values in customer types are:
    ['Critical' 'Medium' 'Not Specified' 'Low' 'High']

5.what are the most common shipping modes? 
--> Most common shipping mode is : "Regular Air"

6.Which cities have the highest number of orders? 
-->

    Sydney       3584
    Melbourne    1416

7.Whats the range of order quantities and prices? 
-->

    Range of Order Quantities is in between 1.0 to 50.0
    Range of Price is in between 0.24 to 377.99

8.Are there any orders with zero or negative total or quantity? 
--> There is no order with zero or negative total or quantity.

9.Are all discount percentages matching discount dollar amounts?
--> All discount percentages are not matching discount dollar amounts.

10.Check for mismatches in total calculation
--> Found missmatches in total 3960 rows.

11.Identify top 5 products by order quantity.

    Product Name
    Artisan 474 Labels                         136
    Smiths General Use 3-Ring Binders          126
    Apex Preferred Stainless Steel Scissors    112
    Artisan 479 Labels                         112
    Alto Memo Cubes                            107

12.Which Account Manager handled the most revenue?
--> 

    Account Manager
    Connor Betts    928249.39

13.What is the average shipping cost by mode?
-->

    Ship Mode
    Regular Air       5.503553
    Delivery Truck    5.270085
    Express Air       5.247311

14.Find the most profitable product.
--> 

    Product Name
    Economy Rollaway Files    $97.47

15.What is the total revenue generated across all orders? 
-->Total revenue generated across all orders is :- 5186314.12

16.Which customer type generates more revenue? 

    Customer Type
    Corporate    1753097.41

17.How does order priority affect revenue? 
--> Correlation is (0.007173550710202885)

18.What is the average profit margin by product category?

    Product Category
    Furniture          12.701953
    Office Supplies    17.682609
    Technology         20.423208

19.What is the most profitable product overall?
--> Most profitable product is :- Emerson Stylus 1520 Color Inkjet Printer

20.How many days does it usually take to ship an order? 
-->Usually it takes 1 day to ship an order.

21.Do longer shipping times impact profit margins?
--> Correlation between shipping time and profit margin is : (-0.0197)

22.Which city brings in the highest revenue? 

    City
    Melbourne    29999.5

23.Which shipping mode is most cost-effective (lowest avg. shipping)?

    Ship Mode
    Express Air    5.247311

24.Do higher discounts reduce profits
--> Correlation between discount and profits is : 0.3934

25.Which state has the highest number of orders? 

    State
    NSW    3583

26.What is the average discount % across all orders? 
--> Average discount % across all orders is 0.050

27.What is the average total spend per order? 
--> Average total spend per order is 746.30

28.Are certain containers (e.g., Small Box, Wrap Bag) more profitable?
--> Most profitable container is Small Box with profit of 19.67%

